### DISCUSSION TITLE
What are the coding errors regarding this Python export code? This is meant to be run in the "Execute Script" window in Python mode.

### DISCUSSION THREAD

**Discussion by CubeSolver25** [fontforge/fontforge#4975] [Q&A] [answered]
I am currently working on an SVG export system which would allow me to export desired characters by specifying the characters I want to export, the font I want to use, export destination, and em (font size).

[Code builds upon the solution found in this post.](https://github.com/fontforge/fontforge/issues/4540)

```py
import fontforge
import psMat
from pathlib import Path
from PIL import Image

#Don't change this.
specialchars = ' <>:"/\|?*.%^'
specialcharnames = ['space', 'lessthan', 'greaterthan', 'colon', 'doublequote', 'forwardslash', 'backslash', 'pipe', 'question', 'asterisk', 'period', 'percent', 'caret']


fontfile = 'Path to font file goes here'
fontfile = fontfile.replace('\\ ', ' ')

destination = 'Path to export destination goes here'
destination = destination.replace('\\ ', ' ')

targetEm = 32

allupperchars = ""

#ASCII 0-127 Characters (Whitespace included):
# !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~
#ASCII 0-255 Characters (Whitespace included):
# !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£¤¥¦§¨©ª«¬®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞ

characters = ' !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'

f = fontforge.open(fontfile)
f.em = targetEm

for i, char in enumerate(characters):
    prefix = ""
    print(char)
    g = f[str(char)]
    if char in specialchars:
        char = specialcharnames[specialchars.index(char)]
        print("Special Character Detected: " + char)
    elif char.isalpha():
        if char.isupper():
            prefix = "upper"
            allupperchars = allupperchars + char    
    g.export(str(Path(destination) / (prefix + char + '.svg')))

```
The main issue is with `g = f[str(char)]`

The expected behavior is that the program would load the glyph based off the current character within the string. 
However, attempting to run this code gives me a **No such glyph** error, even though I am trying to convert the character into a string. 
<img width="624" alt="Screen Shot 2022-03-30 at 4 14 19 PM" src="https://user-images.githubusercontent.com/36614761/160945902-7760f4d1-6b07-4e95-ae2b-f9a3b8c8f03a.png">

Running this code with `g = f['u']` works fine.
<img width="650" alt="Screen Shot 2022-03-30 at 4 07 19 PM" src="https://user-images.githubusercontent.com/36614761/160945273-432c95cd-c3bc-44ed-9c6d-09ccb8ce3eea.png">

Based on the filenames, **char** should be a single character string, and should be able to work within that script.

Is this a programming error, or is it due to a misunderstanding on my part as to how Fontforge works with Python?

**The following parts of the program do work when the character is specified as a string:**
Importing & Opening the File
Exporting to Destination
Choosing EM

**Comment by skef** [comment]
First, try changing `characters` to just the upper and lower case letters and see what happens.

```
characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'
```

**Comment by skef** [comment, marked as answer]
If that works the problem is that the glyphs for ASCII characters aren't generally named 1:1 by those characters. 'a' is 'a' but '9' is 'nine'. The shortest path from what you're doing to what you want is to retrieve based on the unicode codepoint of the character you've extracted, which __getattr__ on a font also accepts. so change

```
    g = f[str(char)]
```

to

```
    g = f[ord(char)]
```

**Reply by CubeSolver25** [reply]
By doing this all the graphics immediately got exported correctly. Thank you!

**Comment by skef** [comment]
(You'll probably also want to put that line within a `try` block, and `continue` if you get an exception, given that you appear to be looking for what glyphs are in the font. And you may or may not want to save the files named as "prefix + g.glyphname + '.svg'" to be a bit more filesystem-friendly.)

**Reply by CubeSolver25** [reply]
I plan to do more with this (like being able to choose a label that goes before the character, as it's going to be used for a text engine. If it helps anyone else who may find a use for it, I will upload the finished code when ready.
